File: I-144-1/BinaryCode.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE 
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE 
typedef long long lint;
typedef vector<string> vs;
#define fora(i,v) for(int i = 0; i < (v).size(); i++)
struct BinaryCode
{
	string calc( string s, string m )
	{
		fora(i, m)
		{
			char c = m[i] - s[i] + '0' - (i ? s[i - 1] - '0' : 0);
			if ( c != '0' && c != '1' )
				return "NONE";
			if ( i < m.size() - 1 ) s += c;
		}
		if ( s[s.size() - 1] + (s.size() > 1 ? s[s.size() - 2] - '0' : 0) != m[m.size() - 1] )
			return "NONE";
		return s;
	}
	vs decode( string message )
	{
		vs result;
		result.push_back( calc( "0", message ) );
		result.push_back( calc( "1", message ) );
		return result;
	}
// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const lint& expected, const lint& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<string> A0; vector<vs> A1;
	{ /*Test 0*/string Arg0 = 
	"123210122"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "011100011",  "NONE" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	{ /*Test 1*/string Arg0 = 
	"11"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "01",  "10" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	{ /*Test 2*/string Arg0 = 
	"22111"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "NONE",  "11001" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	{ /*Test 3*/string Arg0 = 
	"123210120"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "NONE",  "NONE" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	{ /*Test 4*/string Arg0 = 
	"3"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "NONE",  "NONE" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	{ /*Test 5*/string Arg0 = 
	"12221112222221112221111111112221111"
	; A0.push_back( Arg0 ); string _Arg1[] = 
	{ "01101001101101001101001001001101001",  "10110010110110010110010010010110010" }
	; vs Arg1(_Arg1, _Arg1 + (sizeof(_Arg1) / sizeof(_Arg1[0]))); A1.push_back( Arg1 );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A1[i], decode( A0[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE 
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	BinaryCode ___test; 
	___test.run_test(-1); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE 

File: II-144-2/BinaryCode.cs
using System;
using System.Collections;
using System.Text;
using System.Text.RegularExpressions;

public class BinaryCode
{
	private int getDigit( string s, int i )
	{
		if ( i < 0 ) return 0;
		if ( i >= s.Length ) return 0;
		return int.Parse( s[i].ToString() );
	}
    public string[] decode( string message )
    {
		string a = "0";
		string b = "1";
		for( int i = 0; i < message.Length - 1; i++ )
		{
			int x = getDigit( message, i ) - getDigit( a, i - 1 ) - getDigit( a, i );			
			if ( x != 0 && x != 1 )
			{
				a = "NONE";
				break;
			}
			a += x.ToString();
		}
		if ( a != "NONE" && getDigit( message, message.Length - 1 ) != 
			getDigit( a, message.Length - 1 ) + getDigit( a, message.Length - 2 ) )
		{
			a = "NONE";
		}
		for( int i = 0; i < message.Length - 1; i++ )
		{
			int x = getDigit( message, i ) - getDigit( b, i - 1 ) - getDigit( b, i );			
			if ( x != 0 && x != 1 )
			{
				b = "NONE";
				break;
			}
			b += x.ToString();
		}
		if ( b != "NONE" && getDigit( message, message.Length - 1 ) != 
			getDigit( b, message.Length - 1 ) + getDigit( b, message.Length - 2 ) )
		{
			b = "NONE";
		}
		return new string[] {a, b};
    }

// BEGIN CUT HERE
  public static void Main(string[] args)
  {
    try
    {
		eq(1,(new BinaryCode()).decode("123210122"),new String[] { "011100011",  "NONE" });
		eq(2,(new BinaryCode()).decode("11"),new String[] { "01",  "10" });
		eq(3,(new BinaryCode()).decode("22111"),new String[] { "NONE",  "11001" });
		eq(4,(new BinaryCode()).decode("123210120"),new String[] { "NONE",  "NONE" });
		eq(5,(new BinaryCode()).decode("3"),new String[] { "NONE",  "NONE" });
		eq(6,(new BinaryCode()).decode("12221112222221112221111111112221111"),new String[] { "01101001101101001101001001001101001",  "10110010110110010110010010010110010" });
    }
    catch(Exception e)
    {
	Console.WriteLine(e.Message);
    }
  }
  private static void eq(int n, int a, int b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, char a, char b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
  }
  private static void eq(int n, long a, long b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, bool a, bool b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, string a, string b)
  {
    if(a == b)
      Console.WriteLine("Case "+n+" passed.");
    else
      Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
  }
  private static void eq(int n, int[] a, int[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, long[] a, long[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void eq(int n, string[] a, string[] b)
  {
    if(a.Length != b.Length)
    {
      Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
      return;
    }
    for(int i= 0; i < a.Length; i++)
      if(a[i] != b[i])
      {
        Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
        print(b);
        print(a);
        return;
      }
    Console.WriteLine("Case "+n+" passed.");
  }
  private static void print(int a)
  {
    Console.Write(a+" ");
  }
  private static void print(string s)
  {
    Console.Write("\""+s+"\" ");
  }
  private static void print(int[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(long[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write(rs[i]);
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void print(string[] rs)
  {
    if(rs == null) return;
    Console.Write('{');
    for(int i= 0; i < rs.Length; i++)
    {
      Console.Write("\""+rs[i]+"\"");
      if(i != rs.Length-1)
        Console.Write(", ");
    }
    Console.WriteLine('}');
  }
  private static void nl()
  {
    Console.WriteLine();
  }
// END CUT HERE
}
